crml/cennzx/src/impls.rs

```diff
@@ -15,26 +15,58 @@
 //!
 //! Extra CENNZX-Spot traits + implementations
 //!
-use crate::{Module, Trait, MODULE_ID};
+use crate::{Module, Trait};
 use cennznet_primitives::{
 	traits::{BuyFeeAsset, SimpleAssetSystem},
 	types::FeeExchange,
 };
-use frame_support::dispatch::DispatchError;
-use sp_runtime::traits::AccountIdConversion;
+use frame_support::dispatch::{DispatchError, DispatchResult};
+use prml_generic_asset::{AssetIdAuthority, MultiCurrencyAccounting};
+use sp_core::crypto::{UncheckedFrom, UncheckedInto};
+use sp_runtime::traits::Hash;
+use sp_std::{marker::PhantomData, prelude::*};
+
+/// A function that generates an `AccountId` for a CENNZX exchange / (core, asset) pair
+pub trait ExchangeAddressFor {
+	/// The Account Id type
+	type AccountId;
+	/// The Asset Id type
+	type AssetId;
+	/// Create and exchange address given `asset_id`
+	fn exchange_address_for(asset_id: Self::AssetId) -> Self::AccountId;
+}
+
+/// A CENNZX exchange address generator implementation
+pub struct ExchangeAddressGenerator<T: Trait>(PhantomData<T>);
 
-/// Generate a CENNZX exchange address for a core, `asset_id` pair
-pub fn exchange_address_for<T: Trait>(asset_id: T::AssetId) -> T::AccountId {
-	let core_asset_id = Module::<T>::core_asset_id();
-	MODULE_ID.into_sub_account((core_asset_id, asset_id))
+impl<T: Trait> ExchangeAddressFor for ExchangeAddressGenerator<T>
```

Comment:
- ModuleId doesn't provide enough bytes to be used here
- Couldn't find a way to refactor this as a function.
It needs the type awareness provided by associated types when configured in the runtime
to satisfy the trait bounds here

Reply:
8 bytes (ModuleId) could be enough for making unique exchange address, right? Looking at other module ids:
b"phrelect", b"py/trsry", b"py/socie", a name like b"cennz/xs" or b"cennz-xs" or similar could be a good candidate, I'm thinking.
Do we have any backwards compatibility issue here? Like previously generated exchange addresses?

Reply:
ModuleId will give us one address but at the moment we need a unique address for each pool.
We could use into_sub_account but that relies on the encode implementation to guarantee uniqueness and truncates at 32 bytes 😬

so module_id.into_sub_account((core_id, asset_id)) won't take into account the full 128 bytes of entropy in the asset IDs which could lead to the same addresses being used which will throw off the liquidity accounting.

Hashing the bytes seems to be the only way to reasonably guarantee this is unique.